File: get_gps_location.ts
import * as fs from "fs";
import * as path from "path";
import * as os from "os";

interface Location {
  latitude: number;
  longitude: number;
  location_name: string;
}

const LOCATION_FILE = path.join(os.homedir(), ".swifttow_location.json");

function saveLocation(
  latitude: number,
  longitude: number,
  locationName: string = ""
): void {
  const data: Location = {
    latitude,
    longitude,
    location_name: locationName,
  };

  fs.writeFileSync(LOCATION_FILE, JSON.stringify(data, null, 2));
  console.log(`\n✓ Location saved: ${locationName}`);
  console.log(`  Lat: ${latitude}, Lon: ${longitude}\n`);
}

function getLocation(): Location | null {
  if (fs.existsSync(LOCATION_FILE)) {
    const data = fs.readFileSync(LOCATION_FILE, "utf-8");
    return JSON.parse(data);
  }
  return null;
}

function displayLocation(data: Location): void {
  console.log("\n📍 Your Saved Location:");
  console.log("========================================");
  console.log(`Latitude:   ${data.latitude}`);
  console.log(`Longitude:  ${data.longitude}`);
  if (data.location_name) {
    console.log(`Location:   ${data.location_name}`);
  }
  console.log("========================================\n");
}

const args = process.argv.slice(2);

if (args.length > 0) {
  if (args[0] === "set") {
    if (args.length >= 3) {
      const lat = parseFloat(args[1]);
      const lon = parseFloat(args[2]);
      const name = args[3] || "";
      saveLocation(lat, lon, name);
    } else {
      console.log(
        "Usage: ts-node get_gps_location.ts set <latitude> <longitude> [location_name]"
      );
    }
  } else if (args[0] === "get") {
    const data = getLocation();
    if (data) {
      displayLocation(data);
    } else {
      console.log("No location saved. Use: ts-node get_gps_location.ts set <lat> <lon> [name]");
    }
  }
} else {
  const data = getLocation();
  if (data) {
    displayLocation(data);
  } else {
    console.log("No location saved yet.");
    console.log("Save your location with:");
    console.log("  ts-node get_gps_location.ts set <latitude> <longitude> [location_name]");
    console.log("\nExample:");
    console.log("  ts-node get_gps_location.ts set 0.592857 35.352493 Kelji");
  }
}

File: get_gps_location.py
#!/usr/bin/env python3
import json
import os
from pathlib import Path

LOCATION_FILE = os.path.join(os.path.expanduser("~"), ".swifttow_location.json")

def save_location(latitude, longitude, location_name=""):
    """Save current location coordinates"""
    data = {
        "latitude": latitude,
        "longitude": longitude,
        "location_name": location_name
    }
    with open(LOCATION_FILE, 'w') as f:
        json.dump(data, f, indent=2)
    print(f"\nLocation saved: {location_name}")
    print(f"  Lat: {latitude}, Lon: {longitude}\n")

def get_location():
    """Retrieve saved location"""
    if os.path.exists(LOCATION_FILE):
        with open(LOCATION_FILE, 'r') as f:
            data = json.load(f)
        return data
    return None

def display_location(data):
    """Display location nicely"""
    print("\n📍 Your Saved Location:")
    print("=" * 40)
    print(f"Latitude:   {data['latitude']}")
    print(f"Longitude:  {data['longitude']}")
    if data.get('location_name'):
        print(f"Location:   {data['location_name']}")
    print("=" * 40 + "\n")

if __name__ == "__main__":
    import sys
    
    if len(sys.argv) > 1:
        if sys.argv[1] == "set":
            if len(sys.argv) >= 4:
                lat = float(sys.argv[2])
                lon = float(sys.argv[3])
                name = sys.argv[4] if len(sys.argv) > 4 else ""
                save_location(lat, lon, name)
            else:
                print("Usage: python get_gps_location.py set <latitude> <longitude> [location_name]")
        elif sys.argv[1] == "get":
            data = get_location()
            if data:
                display_location(data)
            else:
                print("No location saved. Use: python get_gps_location.py set <lat> <lon> [name]")
    else:
        data = get_location()
        if data:
            display_location(data)
        else:
            print("No location saved yet.")
            print("Save your location with:")
            print("  python get_gps_location.py set <latitude> <longitude> [location_name]")
            print("\nExample:")
            print("  python get_gps_location.py set 0.592857 35.352493 Kelji")
